### PR TITLE
3D visualizations feature

### PR DESCRIPTION
**Pull request by alior101**
Hi Martin,
After lots of work  trying many 3D libs I finally settled on Easy3D lib - I wrapped it in LayersVisualization lib since the original lib was not fit for multithreading opengl working and gpu buffers online update. 
I created a new tim_visualize example in which you can see in real time the activity of neurons with the same cell column structure... This is of course just a starting point and the real benefit will come from visualizing grid cell planes and their multi layers influence ...
Tell me if you have any issue with it ... 
Lior 

Screenshot from 2019-07-16 12-07-58

### CODE REVIEW

**Comment by marty1885** [comment]
@alior101 Wow, impressive work! I'll try to build it tomorrow.

It seems that you have included the entire Easy3D library in Etaler repo. Would you mind moving it into separate repo and add it as a submodule? So we can manage the library easier, and pulling from upstream becomes easier too. If you like to, I can open a repo in Etaler's organization.

**Comment by alior101** [comment]
Now with easy3d as submodule :)

**Comment by marty1885** [comment]
@alior101 Thanks! After further viewing the changed files. It seems that you have also pulled in the entirety of ImGui and (seems to be) Easy3D's shader source code. Please put them in another repository so we can keep up with upstream.

Also, Easy3D is licensed under GPLv3 while Etaler is released under BSD 3 clause. Which aren't compatible with each other... Maybe we could make your visualizer a standalone application/library? Or any suggestion?

**Comment by alior101** [comment]
Hi Martin.
Imgui can be easily converted to submode. I haven't thought about the
licensing issues.. It seems like it's a real issue.. I might make a
separate repo of LayersVisualizer as a lib.. Will that be OK in term of
license if it's submoduled into etaler ?..

**Comment by alior101** [comment]
Also, any preferences as to shared or static lib in term of license
compatability?

**Comment by marty1885** [comment]
> Will that be OK in term of license if it's submoduled into etaler ?..

Since the LayerVisualizer isn't used anywhere is the core library (please confirm this). We could move the LayerVisualizer into it's own directory under the Etaler's root directoy. And add a GPLv3 licence there. Turning it into a standalone yet tightly integrated lib. 

Either static/dynamic is fine here.

**Comment by alior101** [comment]
Hi Martin,
What about holding it in a separate repo and just submodule it inside
tm_visualize example since it's not used elsewhere in the core ? It is just
a simple 3d point and connections cloud visualization lib so maybe other
people could make use of it in visualizing general 3d connected nodes ..
What do you say ? Is it still OK in term of license ?

**Comment by marty1885** [comment]
You mean hold LayerVisualizer in a separate repo and submodule it inside tm_visualize? Yes, that will also work. Let me know if I got things wrong. And please remember add GPLv3 deceleration in the top of tm_visualize.

**Comment by alior101** [comment]
done :)

File: docs/GUI.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+This file documents the steps to add GUI features to Etaler running under container.
+* on the host run "xhost local:root" to enable X11 connection on localhost from the container
+* Enable ETALER_ENABLE_MATPLOTLIB_CPP in the examples directory cmake list file
```
Comment (marty1885) [inline docs/GUI.md]: I believe you are not using matpliotlib anymore. Please update the documents. Let me know if I'm wrong.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,8 @@
+FROM  ubuntu:18.04
+
+RUN rm /bin/sh && ln -s /bin/bash /bin/sh
+RUN DEBIAN_FRONTEND=noninteractive apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y tzdata g++  wget git opencl-headers \
+    ocl-icd-opencl-dev libcereal-dev sudo nano cmake apt-utils \
+    libtbb-dev gdb python python-dev python-matplotlib python-numpy python2.7-dev policykit-1 x11-apps xorg-dev libglu1-mesa-dev libglew-dev libglfw3-dev
```
Comment (marty1885) [inline .devcontainer/Dockerfile]: AFAIK you have removed matplotlib and matplotlib-cpp's usage. Please update

File: examples/CMakeLists.txt [hunk 3]
```diff
@@ -7,16 +7,39 @@ project(spbench CXX)
...
+
+project(tm_visualize CXX)
+link_directories(${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer/build/)
+link_directories(${CMAKE_CURRENT_SOURCE_DIR}/visualize//LayersVisualizer/Easy3D/build/lib)
+include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer/)
+include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer/Easy3D)
+include_directories( ${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer/imgui)
+add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer ${CMAKE_CURRENT_SOURCE_DIR}/visualize/LayersVisualizer/build/)
+find_package(OpenGL REQUIRED)
+SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++17 -pthread -g")
```
Comment (marty1885) [inline examples/CMakeLists.txt]: Please use `set_target_properties(${TARGET} PROPERTIES COMPILE_FLAGS ${ADDITIONAL_BUILD_FLAGS})` to get flags for a single target. Though I believe these flags are set in /CMakeLists.txt already. (You don't need -g and -pthread is provided by tbb)

File: examples/CMakeLists.txt [hunk 4]
```diff
@@ -7,16 +7,39 @@ project(spbench CXX)
...
+add_executable(tm_visualize visualize/tm_visualize.cpp) 
+target_link_libraries(tm_visualize ${OPENGL_LIBRARIES} Etaler LayersVisualizer glew core viewer glfw rply tinyfiledialogs util LAStools fileio dl glew glfw  ${OPENGL_LIBRARIES} X11 )
```
Comment (marty1885) [inline examples/CMakeLists.txt]: Redundant `${OPENGL_LIBRARIES}`. And.. why do we need to link against X11?

File: examples/tmexample.cpp [hunk 5]
```diff
@@ -48,7 +47,26 @@ int main()
 		std::cout << "input, prediction of next = " << categoery
 			<< ", " << to_string(pred_category);
 		std::cout << '\n';
+	/*
```
Comment (marty1885) [inline examples/tmexample.cpp]: Please revert tmexample.cpp. It is a example program for others to learn from

File: docs/GUI.md (hunk 1 above)
Comment (alior101) [reply docs/GUI.md]: NP

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (alior101) [reply .devcontainer/Dockerfile]: NP

File: examples/tmexample.cpp (hunk 5 above)
Comment (alior101) [reply examples/tmexample.cpp]: NP

File: examples/CMakeLists.txt (hunk 3 above)
Comment (alior101) [reply examples/CMakeLists.txt]: NP

File: examples/CMakeLists.txt (hunk 4 above)
Comment (alior101) [reply examples/CMakeLists.txt]: NP.

**Comment by alior101** [comment]
All your reviews are OK. The only issue I have is with the resource dir ... Easy3D internal viewer is opening a hardcoded  "resources/shaders" path dir and if the shaders are not found the GPU can not render points and lines ... I don't want to fork Easy3D for the sole purpose of be able to specify shaders dir, but I have no problem putting them into another separate repo ..

**Comment by alior101** [comment]
Hi Martin,
I moved the resources to the visualize directory ... this means that the tm_visualize will have to be run from the build dir and not from the etaler workspace root since easy3d lib expects to load the resources from ./resources  - this is mostly relevant to vscode development env .. where the cmake plugin runs the example from the etaler root ..

**Comment by marty1885** [comment]
Sorry I was away in the weekends. I'll make some changes so the vosualozer also works with the GPU. And it should be ready to merge.

BTW, seems you have deleted the docker/ folder. Why is that?

**Comment by marty1885** [comment]
@alior101 Also, is the `.gdbinit` file required? If not, please remove it.

**Comment by alior101** [comment]
It's automatic if you work with vscode - if not, there is a docker file but
it's just for building the image. Inside you will have to clone from repo.
I highly recommend working with vscode since it has an automatic remote
development inside a container..  I know you don't have issue with c++17
but many debian based distro are way behind and are missing lots of the
features your code relies upon ..

**Comment by alior101** [comment]
Hi Martin,This gdbinit contains many useful macros for working with stl ..
I know you are working with cling so you don't need it but maybe other will
do. Btw, I like this cling idea and trying to work with it to implement the
grid cells functionality ... It's cool that it has root and it's graphics
capabilities so tightly attached .. I saw it has GL exmaples so maybe it
will replace LayersVissualizer library in the future when I master it
enough ...

**Comment by marty1885** [comment]
About the docker script's location. I was worrying that people who doesn't use VSCode won't know that folder contains docker scripts. And the common place to store Dockerfile is in `/docker`.

And for the .gdbinit file. I was thinking that people might have their own setup and will replace it with their own. Now I'm thinking that weather providing an alternative default is a good thing to do.

**Comment by marty1885** [comment]
@alior101 I think it is good to have easy access of docker enviroment in VSCode. But at the same time, others not using VSCode will have trouble finding the dockerfile. Whould you mind me putting a READE in /docker to redirect people?

**Comment by alior101** [comment]
sure

**Comment by marty1885** [comment]
@alior101 A final problem. In your submodules:

```
[submodule "Etaler/3rdparty/Easy3D"]
	path = Etaler/3rdparty/Easy3D
	url = https://github.com/LiangliangNan/Easy3D.git
	branch = master
[submodule "Etaler/3rdparty/LayersVisualizer/Easy3D"]
	path = Etaler/3rdparty/LayersVisualizer/Easy3D
	url = https://github.com/LiangliangNan/Easy3D.git
```

Why is there 2 instance of Easy3D? Are they both needed?

**Comment by alior101** [comment]
sorry.. :) fixed

**Comment by marty1885** [comment]
Looks good to me! Merging.

**Pull request merged**